pkg/provider/terraform/instance/apply.go

```diff
@@ -329,6 +353,146 @@ func (p *plugin) handleFiles(fns tfFuncs) error {
+	resType TResourceType,
+	resName TResourceName,
+	filename string,
+	props TResourceProperties,
+) {
+	resNameFileProps, has := m[resType]
+	if !has {
+		resNameFileProps = make(map[TResourceName]TResourceFilenameProps)
+		m[resType] = resNameFileProps
+	}
+	resNameFileProps[resName] = TResourceFilenameProps{
+		FileName:  filename,
+		FileProps: props,
+	}
+}
+
+// handleFilePruning processes the prune file that are no longer associated with resources
+// in terraform. If the resource was not in terraform state before the refresh then
+// the backend cloud is queried in order to determine if it still exists; if it does then
+// it is imported and, if it does not, then the file is pruned.
+func (p *plugin) handleFilePruning(
+	fns tfFuncs,
+	prunes map[TResourceType]map[TResourceName]TResourceFilenameProps,
+	tfStateBeforeRefresh map[TResourceType]map[TResourceName]struct{},
+) error {
+	// Determine files to prune, since multiple resource types can exist per file we want to
+	// track unique filenames
+	pruneFiles := make(map[string]struct{})
+	// If the resource was removed out-of-band then it had a previous entry in the state file
+	// and can be pruned; if there is no entry then query the backend to determine if the
+	// resource still exists
+	for resType, resNameFilenameProps := range prunes {
+		var tfResTypeNameProps map[TResourceName]struct{}
+		tfResTypeNameProps, has := tfStateBeforeRefresh[resType]
+		if !has {
+			tfResTypeNameProps = make(map[TResourceName]struct{})
+		}
+		for resName, resFilenameProps := range resNameFilenameProps {
+			if _, has := tfResTypeNameProps[resName]; has {
+				log.Infof("Pruning %v file, resource %v.%v previously existed in terraform",
+					resFilenameProps.FileName,
+					resType,
+					resName)
+				pruneFiles[resFilenameProps.FileName] = struct{}{}
+			} else {
+				// Find resource type in backend
+				importID, err := fns.getExistingResource(resType, resName, resFilenameProps.FileProps)
+				if err != nil {
+					return err
+				}
+				// No ID returned, prune file
+				if importID == nil {
+					log.Infof("Pruning %v file, resource %v.%v was not found in backend",
+						resFilenameProps.FileName,
+						resType,
+						resName)
+					pruneFiles[resFilenameProps.FileName] = struct{}{}
+				} else {
+					// Import resource
+					log.Infof("Importing %v %v into terraform as resource %v ...",
+						string(resType),
+						*importID,
+						string(resName))
+					if err = fns.tfImport(resType, string(resName), *importID); err != nil {
+						return err
+					}
+				}
+			}
+		}
+	}
+	log.Infof("Pruning %v tf.json files", len(prunes))
+	for file := range pruneFiles {
+		path := filepath.Join(p.Dir, file)
+		log.Infof("Pruning file: %v", file)
+		err := p.fs.Remove(path)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// getExistingResource queries the backend cloud to get the ID of the resource associated
+// with the given type, name, and properties
```

Comment:
As discussed in #783 do you think storing the state of an update will obviate this - at least for the rolling update?   Presumably the manager will never fail over unexpectedly if we implemented safe ordering of managers to update and there will be state in raft that the new leader can read and resume.

Reply:
Also if solutions for #783 is implemented then the case 2 as you describe would  never happen, right?  At least not intentionally...

Reply:
@chungers Yes, you are correct. If we failover to a new manager _before_ we start the rolling update then this should not be needed.

However, this PR fixes the generic case when the container dies/restarts _during_ a provision for any reason; for example, if Infrakit is deployed as a Docker service, then doing something like `docker service update --image <new-image>` would cause the associated containers to restart.